### DISCUSSION TITLE
Transitions on appending and removing elements [working prototype]

### DISCUSSION THREAD

**Discussion by mtk3d** [livewire/livewire#4702] [3. Feature ideas]
Hi, I'm started learning how to use Livewire, and something that surprised me was that there are no DOM transitions out of the box.

I made research, and I found an article on Caleb's blog that describes `wire:transition` in probably one of the alpha versions of Livewire. In released versions, it was removed. I also found that someone tried to create a pull request with transitioning, but it didn't handle some of the cases.

I also made some research on how the transitioning is implemented in a few popular frontend frameworks. One of them especially interested me. It's transitioning implementation in Svelte.

Svelte is using CSS `@keyframes` animations to animate elements transitions but the keyframes are generated dynamically using simple functions that are calculates CSS properties for every frame of the animation. I think it may be a powerful solution. My first thought was that it's also possible to generate keyframes directly from PHP, but now I think it would require to pass computed styles of animation element from frontend to backend, where are 71 properties 😄  
But anyway maybe not using PHP functions, but just JS, why not?

To better investigate how those animations work I prepared a simple demo of using them: https://codepen.io/mtk3d/pen/GROQxrJ
Also, the animations here can be stopped and replaced with another one before it ends. Everything is calculated using JavaScript, so we know when the animation is started and finished.

What do you think about implementation like this in Livewire?

**Comment by mtk3d** [comment]
I also prepared a demo of a transition that's working in both directions, using a single animation function: https://codepen.io/mtk3d/pen/abVqKjE
Code is not prettiest, but I've got how it works 😄

**Comment by mtk3d** [comment]
I prepared a draft of my idea if someone would be interested in https://github.com/mtk3d/livewire/pull/1/files 😄 

I have a problem with interrupting. I prepared function that creates new animation and returns callback that is executed after animation ends or if it's interrupted for ex. by another animation:
```JavaScript
handleAnimation(node, LEAVE, animationInterruption).then(() => { parentNode.removeChild(node); });

handleAnimation(node, ENTER, animationInterruption);
```
I can also handle interruption
```JavaScript
handleAnimationInterruption(node);
```

Everything works fine for entire transitions in both directions, and when `enter` transition is interrupted, and `leave` transition is started.
The problem is when I'm trying to interrupt the `leave` transition and start `enter` transition. In my solution, the `removeChild` function in morphdom is executed after the animation ends or when it's interrupted. The interruption happens at the start of the morphing if needed, but it's too late because `fromNode` is already loaded, and `parentNode.removeChild(node);` do not affect it, so `fromNode` still keep an element that should be removed and morphing is not running the appending of the element, so the animation is not started.

I know that morphdom maybe not be the best place for running those handlers but I didn't found yet how to handle element removing to keep it on transition time in DOM and remove after in other way.

**Comment by mtk3d** [comment]
Ok, finally after a long time of searching I found the commit where transitions were removed 😄  https://github.com/livewire/livewire/commit/5707bd6d39f2d61f79906ab07371d669aafb2fb1
I hoped that this commit will tell me more about that, but unfortunately no 😞 
@calebporzio could you reveal in more detail why it was removed from livewire?

**Comment by mtk3d** [comment]
Ok, finally with help of an old implementation I have a working prototype, and it's pushed to the branch on my [fork](https://github.com/mtk3d/livewire/pull/1/files).

![ezgif com-gif-maker](https://user-images.githubusercontent.com/12657846/155158918-f52ea603-d5a7-457a-8ec4-edaa5495d303.gif)

I think the API may look like this:
```HTML
<div wire:transition.linear.fly="{ x: 100, d: 2000 }">
    Example text
</div>
```
Where `linear` is timing function `fly` is an animation type. The rest is specific for animation type, in this case, `x` is vertical movement, and `d` is the duration.

/* EDIT START */
After consideration maybe this one will be better:

```HTML
<div wire:transition.linear.fly.2000ms="100px, 0">
    Example text
</div>
```
Just don't know yet how to put the delay value here.
/* EDIT END */

Of course with more types of transitions, and also the option to write a custom one.

There are two bigger problems, with removing used keyframes and with calculating duration if you interrupt animation a few times. It's not hard to fix, but I don't want to spend more time on that if no one will be interested.

Anyway, I think I'm ready to try to implement that in the right way in Livewire. If you're interested in this feature, or you see problems that I can't see, please share your thoughts 😃

**Comment by mtk3d** [comment]
Maybe @calebporzio, @joshhanley, @nuernbergerA, anyone, anything, any reaction? 😄

**Comment by Bowero** [comment]
I like what you are doing, but isn't Alpine.js better suited to take care of this?

**Reply by mtk3d** [reply]
Thanks for the question
In my opinion yes, and no 😄. I like to use Alpine.js, but in this specific case, I think it generates some not necessarily needed code. When I built my first application I was a bit confused about how to create elements with the transition. Combining Alpine and Livewire was not obvious for me in this case. I think transitions are a basic thing, and it will be great to have them out of the box. 
When I built the modal it was natural for me to create the window and put it inside the blade `@if` statement. Then I had to change my solution and only hide the window using `x-show` to get the transition and find how to pass a variable to Alpine. I also have a problem with implementing an in-out transition on the list of elements for example. That requires even more amount of Alpine.js code.

I think some people don't want to care about implementing transitions. They just want to make the app feel "smooth", nothing more. I think so while reading some other discussions about transitions in livewire. Using Alpine.js requires changing some code and adopting the new solution.

Those are my thoughts and hypothesis 😄

**Comment by joshhanley** [comment]
@mtk3d thanks for all your work on this!

I believe the transition system was removed from Livewire by Caleb in favour of using Alpine's transition system.

But that's not to say Livewire shouldn't have it, someone else tried to PR this in #4167. Caleb does say in that PR that Livewire might be able to have transitions back.

I think it would be better to wait until V3 to PR this though, as V3 will be a full re-write of Livewire's font end to make use of a lot of Alpine's code. So in theory this might be able to be implemented using Alpine's transition system which then means we're not trying to maintain two transition implementations (Alpine's and Livewire's).

What do you think?

**Reply by mtk3d** [reply]
I think it would be awesome 😄 
Nice to hear it. Great thanks, for your answer.

**Reply by joshhanley** [reply]
No worries! 🙂
